Objects/object.c

```diff
@@ -2083,9 +2083,23 @@ _dir_locals(void)
     PyObject *names;
     PyObject *locals;
 
-    locals = _PyEval_GetFrameLocals();
-    if (locals == NULL)
+    if (_PyEval_GetFrame() != NULL) {
+        locals = _PyEval_GetFrameLocals();
+    }
+    PyThreadState *tstate = _PyThreadState_GET();
+    locals = _PyEval_GetGlobalsFromRunningMain(tstate);
```

Comment:
`_PyEval_GetFrameLocals()` returns a new reference (leak), which is overwritten by `_PyEval_GetGlobalsFromRunningMain()`.  So, the fallback looks to be missing.

The same goes for `builtin_vars()`.

Reply:
fixed